doc/writing-ppxs.mld

```diff
@@ -0,0 +1,519 @@
+
+{@ocaml[
+  # let context = Extension.Context.expression;;
+  val context : expression Extension.Context.t =
+    Ppxlib.Extension.Context.Expression
+]}
+
+{3 The Extender Name}
+
+The extension point name on which it applies is simply a string.
+
+{@ocaml[
+  # let extender_name = "add_suffix" ;;
+  val extender_name : string = "add_suffix"
+]}
+
+See below for examples on when the above name and context will trigger rewriting:
+{@ocaml[
+  (* will trigger rewriting: *)
+  let _ = [%add_suffix "payload"]
+
+  (* won't trigger rewriting: *)
+  let _ = [%other_name "payload"] (* wrong name *)
+  let _ = match () with [%add_suffix "payload"] -> () (* wrong context *)
+]}
+
+{3 The Payload Extraction}
+
+An extension node contains a {{!Ppxlib.Parsetree.payload}[payload]}, which will be passed to the transformation function. However, while this payload contains all information, it is not always structured the best way for the transformation function. For instance, in [[%add_suffix "payload"]], the string ["payload"] is encoded as a structure item consisting of an expression’s evaluation, a constant that is a string.
+
+[ppxlib] allows separating the transformation function from the extraction of the payload’s relevant information. As explained in depth in the {{!"matching-code"}Destructing AST nodes} chapter, this extraction is done by destructing the payload’s structure (which is therefore restricted: [[%add_suffix 12]] would be refused by the rewriter of the example below). The extraction is defined by a value of type
+{{!Ppxlib.Ast_pattern.t}[Ast_pattern.t]}. The {{!Ppxlib.Ast_pattern}[Ast_pattern]} module provides some kind of pattern-matching on AST nodes: a way to structurally extract values from an AST
+node in order to generate a value of another kind.
+
+For instance, a value of type
+[(payload, int -> float -> expression, expression) Ast_pattern.t] means that it
+defines a way to extract an [int] and a [float] from a {{!Ppxlib.Parsetree.payload}[payload]},
+which should be then combined to define a value of type {{!Ppxlib.Parsetree.expression}[expression]}.
+
+In our case, the matched value will always be a {{!Ppxlib.Parsetree.payload}[payload]}, as that's the type for extension points' payloads. The type of the
+produced node will have to match the {{!ext_context}type of extension node we rewrite}, {{!Ppxlib.Parsetree.expression}[expression]} in our example.
+
+{@ocaml[
+  # let extracter () = Ast_pattern.(single_expr_payload (estring __)) ;;
+  val extracter : unit -> (payload, string -> 'a, 'a) Ast_pattern.t = <fun>
+]}
+
+The unit argument is not important.  It is added so that {{:https://v2.ocaml.org/manual/polymorphism.html#ss:valuerestriction}value restriction} does not add noise to the type variables. See the {{!"matching-code"}Destructing AST nodes} chapter and the {{!Ppxlib.Ast_pattern}[Ast_pattern] API} for more explanation on the types and functions used above.
+
```

Comment:
Do you think it is better to have it here rather than in the "destructing code" chapter?

This chapter is only for the "bureaucracy" to register PPXs. And, it alreay includes "single_expr_payload"!

Reply:
I could add such a section in "destructing code" chapter, and make a direct link from here to there. What do you think?

Reply:
I think either would be fine. It will probably mostly be used as a take-it-as-is example, in which case it's part of the boiler-plate. However, people can also take it as a how-to-guide to construct their own pattern. So, maybe the beginning of the destruct section would be a good place. What do you think?

Reply:
I added too many examples to fit them in the "boilerplate", so I added them to the destructing section with a link from the boilerplate to them (for "ready-to-use" patterns).

I prefer to put the example at the end of the destructing section: otherwise, the type does not make sense for someone who has not read the explanation that comes later. Experienced user who only want to quickly refer to examples can use the navigation bar.